abcix/adapter/adapter.go

```diff
@@ -32,59 +35,66 @@ func (app *adaptedApp) OriginalApp() abci.Application {
 	return app.abciApp
 }
 
-func (app *adaptedApp) Info(req abcix.RequestInfo) (resp abcix.ResponseInfo) {
-	abciReq := abci.RequestInfo{}
-	if err := copier.Copy(&abciReq, &req); err != nil {
+var typeRegistry = map[string]reflect.Type{
+	"abci.RequestInfo":               reflect.TypeOf(abci.RequestInfo{}),
+	"abci.RequestSetOption":          reflect.TypeOf(abci.RequestSetOption{}),
+	"abci.RequestQuery":              reflect.TypeOf(abci.RequestQuery{}),
+	"abci.RequestCheckTx":            reflect.TypeOf(abci.RequestCheckTx{}),
+	"abci.RequestInitChain":          reflect.TypeOf(abci.RequestInitChain{}),
+	"abci.RequestListSnapshots":      reflect.TypeOf(abci.RequestListSnapshots{}),
+	"abci.RequestOfferSnapshot":      reflect.TypeOf(abci.RequestOfferSnapshot{}),
+	"abci.RequestLoadSnapshotChunk":  reflect.TypeOf(abci.RequestOfferSnapshot{}),
+	"abci.RequestApplySnapshotChunk": reflect.TypeOf(abci.RequestApplySnapshotChunk{}),
+	"abci.RequestEndBlock":           reflect.TypeOf(abci.RequestEndBlock{}),
+	"abci.RequestBeginBlock":         reflect.TypeOf(abci.RequestBeginBlock{}),
+}
+
+func Apply(f interface{}, args ...interface{}) reflect.Value {
+	fun := reflect.ValueOf(f)
+	in := make([]reflect.Value, len(args))
+	for k, arg := range args {
+		in[k] = reflect.ValueOf(arg).Elem()
+	}
+	r := fun.Call(in)
+	return r[0]
+}
+
+func adaptGeneralization(req interface{}, resp interface{}, abciReq string, f interface{}) error {
```

Comment:
nit: naming, how about just calling it `convert`.

Reply:
can you pass in `f` as `func(interface{}) interface{}` then we won't need `Apply` anymore?

Reply:
It seems that once we have specified the input type of a function, the inputs could not be simply represented as `interface{}`.

Reply:
nit: naming, `applyLegacyABCI`